File: app/src/main/java/com/o4x/musical/ui/fragments/mainactivity/home/HomeFragment.java
package com.o4x.musical.ui.fragments.mainactivity.home;

import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.LinearSmoothScroller;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.h6ah4i.android.widget.advrecyclerview.animator.GeneralItemAnimator;
import com.h6ah4i.android.widget.advrecyclerview.animator.RefactoredDefaultItemAnimator;
import com.kabouzeid.appthemehelper.ThemeStore;
import com.kabouzeid.appthemehelper.util.ColorUtil;
import com.kabouzeid.appthemehelper.util.ToolbarContentTintHelper;
import com.o4x.musical.R;
import com.o4x.musical.dialogs.CreatePlaylistDialog;
import com.o4x.musical.imageloader.glide.SongGlideRequest;
import com.o4x.musical.helper.MusicPlayerRemote;
import com.o4x.musical.interfaces.MusicServiceEventListener;
import com.o4x.musical.loader.LastAddedLoader;
import com.o4x.musical.loader.SongLoader;
import com.o4x.musical.loader.TopAndRecentlyPlayedTracksLoader;
import com.o4x.musical.model.Song;
import com.o4x.musical.model.smartplaylist.HistoryPlaylist;
import com.o4x.musical.model.smartplaylist.LastAddedPlaylist;
import com.o4x.musical.ui.activities.MainActivity;
import com.o4x.musical.ui.activities.SearchActivity;
import com.o4x.musical.ui.activities.base.AbsMusicServiceActivity;
import com.o4x.musical.ui.adapter.home.HomeAdapter;
import com.o4x.musical.ui.fragments.mainactivity.AbsMainActivityFragment;
import com.o4x.musical.util.NavigationUtil;
import com.o4x.musical.util.PhonographColorUtil;
import com.o4x.musical.util.Util;
import com.o4x.musical.util.ViewUtil;
import com.xw.repo.widget.BounceScrollView;

import java.util.concurrent.atomic.AtomicBoolean;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class HomeFragment extends AbsMainActivityFragment implements MainActivity.MainActivityFragmentCallbacks {

    private MainActivity activity;

    private Unbinder unbinder;

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.appbar)
    AppBarLayout appbar;
    @BindView(android.R.id.empty)
    View empty;

    @BindView(R.id.header)
    View header;
    @BindView(R.id.poster_parent)
    FrameLayout posterParent;
    @BindView(R.id.poster)
    ImageView poster;
    @BindView(R.id.poster_gradient)
    View posterGradient;

    @BindView(R.id.nested_scroll_view)
    BounceScrollView bounceScrollView;
    @BindView(R.id.queue_recycler_view)
    RecyclerView queueView;
    @BindView(R.id.recently_recycler_view)
    RecyclerView recentlyView;
    @BindView(R.id.new_recycler_view)
    RecyclerView newView;

    @BindView(R.id.shuffle_btn)
    FloatingActionButton shuffleBtn;

    @BindView(R.id.queue_parent)
    ConstraintLayout queueParent;
    @BindView(R.id.open_queue_button)
    Button openQueueButton;
    @BindView(R.id.recently_parent)
    RelativeLayout recentlyParent;
    @BindView(R.id.newly_parent)
    RelativeLayout newlyParent;

    private HomeAdapter queueAdapter;
    private LinearLayoutManager queueLayoutManager;
    private QueueListener queueListener;

    private HomeAdapter recentlyAdapter, newAdapter;
    private GridLayoutManager recentlyLayoutManager, newLayoutManager;

    public static HomeFragment newInstance() { return new HomeFragment(); }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            activity = getMainActivity();
        } catch (ClassCastException e) {
            throw new RuntimeException(context.getClass().getSimpleName() + " must be an instance of " + AbsMusicServiceActivity.class.getSimpleName());
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        activity = null;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home, container, false);
        unbinder = ButterKnife.bind(this, view);

        return view;
    }

    @Override
    public void onDestroyView() {
        activity.removeMusicServiceEventListener(queueListener);

        queueAdapter = null;
        queueLayoutManager = null;
        queueListener = null;

        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        queueListener = new QueueListener();
        activity.addMusicServiceEventListener(queueListener);

        activity.setStatusBarColor(transparentColor());
        activity.setNavigationBarColorAuto();
        activity.setTaskDescriptionColorAuto();

        setUpToolbar();

        setUpViews();
    }

    private int transparentColor() {
        return ColorUtil.withAlpha(ThemeStore.primaryColor(activity), 0);
    }

    private void setUpToolbar() {
        final int transparentColor = transparentColor();
        appbar.setBackgroundColor(transparentColor);
        toolbar.setBackgroundColor(transparentColor);
        toolbar.setNavigationIcon(R.drawable.ic_menu_white_24dp);
        activity.setTitle(R.string.app_name);
        activity.setSupportActionBar(toolbar);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_main, menu);
        if (activity == null) return;
        ToolbarContentTintHelper.handleOnCreateOptionsMenu(activity, toolbar, menu, ThemeStore.primaryColor(activity));
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        if (activity == null) return;
        ToolbarContentTintHelper.handleOnPrepareOptionsMenu(activity, toolbar);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_shuffle_all:
                MusicPlayerRemote.openAndShuffleQueue(SongLoader.getAllSongs(activity), true);
                return true;
            case R.id.action_new_playlist:
                CreatePlaylistDialog.create().show(getChildFragmentManager(), "CREATE_PLAYLIST");
                return true;
            case R.id.action_search:
                startActivity(new Intent(activity, SearchActivity.class));
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean handleBackPress() {
        return false;
    }

    public void setAppbarColor(int color) {
        int colorFrom = ViewUtil.getViewBackgroundColor(appbar);
        int colorTo = color;
        if (colorFrom == colorTo) return;
        ValueAnimator colorAnimation = ValueAnimator.ofObject(new ArgbEvaluator(), colorFrom, colorTo);
        colorAnimation.setDuration(400); // milliseconds
        colorAnimation.addUpdateListener(
                animator -> {
                    int background = (int) animator.getAnimatedValue();
                    appbar.setBackgroundColor(background);
                    toolbar.setBackgroundColor(background);
                }
        );
        colorAnimation.start();
    }

    private void setUpViews() {
        setUpHeights();
        setUpOnClicks();
        setUpBounceScrollView();
        setUpQueueView();
        setUpRecentlyView();
        setUpNewView();
        checkIsEmpty();
    }


    private void setUpHeights() {
        int displayHeight = Resources.getSystem().getDisplayMetrics().heightPixels;
        ViewGroup.LayoutParams params;

        // Set up header height
        params = header.getLayoutParams();
        params.height = displayHeight / 3;
        header.setLayoutParams(params);

        // Set up poster image height
        params = poster.getLayoutParams();
        params.height = (int) (displayHeight / 1.5f);
        poster.setLayoutParams(params);

        // Set up posterGradient height
        posterGradient.setLayoutParams(params);

        // Set up posterGradient gradient
        //create a new gradient color
        int[] colors = {
                Color.TRANSPARENT , Color.TRANSPARENT, Color.TRANSPARENT, Color.TRANSPARENT,
                PhonographColorUtil.getWindowColor(activity)};
        GradientDrawable gd = new GradientDrawable(
                GradientDrawable.Orientation.TOP_BOTTOM, colors);

        posterGradient.setBackground(gd);
    }

    private  void setUpOnClicks() {
        queueParent.setOnClickListener(view -> {
            activity.setMusicChooser(R.id.nav_queue);
        });
        openQueueButton.setOnClickListener(view -> {
            activity.setMusicChooser(R.id.nav_queue);
        });
        recentlyParent.setOnClickListener(view -> {
            NavigationUtil.goToPlaylist(activity, new HistoryPlaylist(activity));
        });
        newlyParent.setOnClickListener(view -> {
            NavigationUtil.goToPlaylist(activity, new LastAddedPlaylist(activity));
        });
        shuffleBtn.setOnClickListener(view -> {
            MusicPlayerRemote.openAndShuffleQueue(SongLoader.getAllSongs(activity), true);
        });
    }

    private void setUpBounceScrollView() {
        final int displayHeight = Resources.getSystem().getDisplayMetrics().heightPixels;
        final int statusBarHeight = Util.getStatusBarHeight(activity);
        final int appbarHeight = toolbar.getLayoutParams().height;


        // get real header height
        final float headerHeight = header.getLayoutParams().height - appbarHeight - statusBarHeight;

        final int transparentColor = transparentColor();
        AtomicBoolean isStatusFlat = new AtomicBoolean(false);
        AtomicBoolean isAppbarFlat = new AtomicBoolean(false);

        bounceScrollView.setOnScrollChangeListener(
                (BounceScrollView.OnScrollChangeListener) (v, scrollX, scrollY, oldScrollX, oldScrollY) -> {

                    // Scroll poster
                    posterParent.setY(
                            // for quality cast to int
                            (int) (-scrollY / (displayHeight * 2 / (float) poster.getLayoutParams().height))
                    );

                    // Scroll appbar
                    if (scrollY > headerHeight + appbarHeight && !isAppbarFlat.get()) {
                        setAppbarColor(ThemeStore.primaryColor(activity));
                        appbar.setElevation(8);
                        isAppbarFlat.set(true);
                    }
                    if (scrollY > headerHeight) {
                        if (!isStatusFlat.get()) {
                            activity.setStatusBarColorAuto();
                            isStatusFlat.set(true);
                        }
                        if (scrollY > oldScrollY) {
                            appbar.setY(
                                    Math.max(
                                            -appbarHeight, appbar.getY() + (oldScrollY - scrollY)
                                    )
                            );
                            shuffleBtn.hide();
                        } else {
                            appbar.setY(
                                    Math.min(
                                            0, appbar.getY() + (oldScrollY - scrollY)
                                    )
                            );
                            shuffleBtn.show();
                        }
                    } else {
                        if (isStatusFlat.get()) {
                            setAppbarColor(transparentColor);
                            activity.setStatusBarColor(transparentColor);
                            appbar.setElevation(0);
                            isStatusFlat.set(false);
                            isAppbarFlat.set(false);
                        }
                        appbar.setY(0);
                    }
                }
        );


        // zooming poster in over scroll
        final ViewGroup.LayoutParams params = poster.getLayoutParams();
        final int width = params.width;
        final int height = params.height;
        bounceScrollView.setOnOverScrollListener(
                (fromStart, overScrolledDistance) -> {
                    final float scale = 1 + (overScrolledDistance / (float) displayHeight);
                    final ViewGroup.LayoutParams mParams =
                            new FrameLayout.LayoutParams(width, (int) (height * scale));
                    poster.setLayoutParams(mParams);
                    posterParent.setLayoutParams(mParams);
                    posterGradient.setLayoutParams(mParams);
                }
        );
    }

    private void setUpQueueView() {
        final GeneralItemAnimator animator = new RefactoredDefaultItemAnimator();
        queueLayoutManager = getLinearLayoutManager();
        queueView.setLayoutManager(queueLayoutManager);
        queueAdapter = new HomeAdapter(
                activity,
                MusicPlayerRemote.getPlayingQueue(),
                MusicPlayerRemote.getPosition(),
                R.layout.item_card_home,
                null,
                false,
                true);
        queueView.setAdapter(queueAdapter);
        queueView.setItemAnimator(animator);
        queueLayoutManager.scrollToPositionWithOffset(MusicPlayerRemote.getPosition(), 0);
    }

    private void setUpRecentlyView() {
        recentlyLayoutManager = getGridLayoutManager();
        recentlyView.setLayoutManager(recentlyLayoutManager);
        recentlyAdapter = new HomeAdapter(
                activity,
                TopAndRecentlyPlayedTracksLoader.getRecentlyPlayedTracks(activity),
                0,
                R.layout.item_card_home,
                getGridSize() * 2,
                false,
                false);
        recentlyView.setAdapter(recentlyAdapter);
    }

    private void setUpNewView() {
        newLayoutManager = getGridLayoutManager();
        newView.setLayoutManager(newLayoutManager);
        newAdapter = new HomeAdapter(
                activity,
                LastAddedLoader.getLastAddedSongs(activity),
                0,
                R.layout.item_card_home,
                getGridSize() * 3,
                false,
                false);
        newView.setAdapter(newAdapter);
    }

    private GridLayoutManager getGridLayoutManager() {
        final int size = getGridSize();
        return new GridLayoutManager(activity, size) {
            @Override
            public boolean checkLayoutParams(RecyclerView.LayoutParams lp) {
                lp.width = getWidth() / size;
                lp.height = (int) (lp.width * 1.5);
                return super.checkLayoutParams(lp);
            }
        };
    }

    private LinearLayoutManager getLinearLayoutManager() {
        final int size = getGridSize();
        return new LinearLayoutManager(activity, LinearLayoutManager.HORIZONTAL, false) {
            @Override
            public boolean checkLayoutParams(RecyclerView.LayoutParams lp) {
                lp.width = getWidth() / size;
                lp.height = (int) (lp.width * 1.5);
                return super.checkLayoutParams(lp);
            }
        };
    }

    private int getGridSize() {
        return getResources().getInteger(R.integer.home_grid_columns);
    }

    class QueueListener implements MusicServiceEventListener {

        QueueListener() {
            updatePoster();
        }

        @Override
        public void onServiceConnected() {
            updatePoster();
            updateQueue();
            checkIsEmpty();
        }

        @Override
        public void onServiceDisconnected() {

        }

        @Override
        public void onQueueChanged() {
            updateQueue();
        }

        @Override
        public void onPlayingMetaChanged() {
            updatePoster();
            updateQueue();
        }

        @Override
        public void onPlayStateChanged() {
            updateQueue();
        }

        @Override
        public void onRepeatModeChanged() {

        }

        @Override
        public void onShuffleModeChanged() {

        }

        @Override
        public void onMediaStoreChanged() {
            updateQueue();
        }

        private void updatePoster() {
            if (!MusicPlayerRemote.getPlayingQueue().isEmpty()) {
                Song song = MusicPlayerRemote.getCurrentSong();
//                if (navigationDrawerHeader == null) {
//                    navigationDrawerHeader = navigationView.inflateHeaderView(R.layout.navigation_drawer_header);
//                    //noinspection ConstantConditions
//                    navigationDrawerHeader.setOnClickListener(v -> {
//                        drawerLayout.closeDrawers();
//                        if (getPanelState() == SlidingUpPanelLayout.PanelState.COLLAPSED) {
//                            expandPanel();
//                        }
//                    });
//                }
//                ((TextView) navigationDrawerHeader.findViewById(R.id.title)).setText(song.title);
//                ((TextView) navigationDrawerHeader.findViewById(R.id.text)).setText(MusicUtil.getSongInfoString(song));
                SongGlideRequest.Builder.from(Glide.with(activity), song)
                        .build()
                        .into(poster);
            } else {
//                if (navigationDrawerHeader != null) {
//                    navigationView.removeHeaderView(navigationDrawerHeader);
//                    navigationDrawerHeader = null;
//                }
            }
        }

        private void updateQueue() {
            queueAdapter.swapDataSet(MusicPlayerRemote.getPlayingQueue(), MusicPlayerRemote.getPosition());
            resetToCurrentPosition();
        }

        private void resetToCurrentPosition() {
            if (queueAdapter.getItemCount() == 0) return;
            queueView.stopScroll();

            final int from = queueLayoutManager.findFirstVisibleItemPosition();
            final int to = MusicPlayerRemote.getPosition();
            final int delta = Math.abs(to - from);

            final int limit = 150;
            if (delta > limit) {
                queueLayoutManager.scrollToPosition(
                        to + (to > from ? -limit : limit)
                );
            }


            RecyclerView.SmoothScroller smoothScroller = new LinearSmoothScroller(activity) {
                @Override
                protected int getHorizontalSnapPreference() {
                    return LinearSmoothScroller.SNAP_TO_ANY;
                }

                @Override
                protected float calculateSpeedPerPixel(DisplayMetrics displayMetrics) {
                    if (delta < 20) {
                        return super.calculateSpeedPerPixel(displayMetrics) * 5;
                    } else {
                        return super.calculateSpeedPerPixel(displayMetrics);
                    }
                }


            };
            smoothScroller.setTargetPosition(to);
            queueLayoutManager.startSmoothScroll(smoothScroller);
        }
    }

    private void checkIsEmpty() {
        if (empty != null) {
            empty.setVisibility(
                    (queueAdapter == null || queueAdapter.getItemCount() == 0) &
                    (recentlyAdapter == null || recentlyAdapter.getItemCount() == 0) &
                    (newAdapter == null || newAdapter.getItemCount() == 0)
                            ? View.VISIBLE : View.GONE
            );
        }
    }

}

File: app/src/main/java/com/o4x/musical/ui/fragments/mainactivity/home/HomeFragment.kt
package com.o4x.musical.ui.fragments.mainactivity.home

import android.app.Activity
import android.content.Intent
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Bundle
import android.view.*
import android.widget.FrameLayout
import androidx.core.net.toFile
import androidx.core.view.isVisible
import androidx.core.widget.NestedScrollView
import androidx.lifecycle.lifecycleScope
import androidx.palette.graphics.Palette
import androidx.recyclerview.widget.GridLayoutManager
import code.name.monkey.appthemehelper.extensions.textColorTertiary
import code.name.monkey.appthemehelper.util.ColorUtil
import code.name.monkey.appthemehelper.util.ToolbarContentTintHelper
import com.bumptech.glide.load.resource.bitmap.BitmapTransitionOptions
import com.github.dhaval2404.imagepicker.ImagePicker
import com.o4x.musical.App
import com.o4x.musical.R
import com.o4x.musical.databinding.FragmentHomeBinding
import com.o4x.musical.extensions.toPlaylistDetail
import com.o4x.musical.helper.GridHelper
import com.o4x.musical.helper.MyPalette
import com.o4x.musical.helper.homeGridSize
import com.o4x.musical.imageloader.glide.module.GlideApp
import com.o4x.musical.model.smartplaylist.HistoryPlaylist
import com.o4x.musical.model.smartplaylist.LastAddedPlaylist
import com.o4x.musical.prefs.HomeHeaderPref
import com.o4x.musical.prefs.PreferenceUtil
import com.o4x.musical.ui.activities.MusicPickerActivity
import com.o4x.musical.ui.adapter.home.HomeAdapter
import com.o4x.musical.ui.dialogs.CreatePlaylistDialog
import com.o4x.musical.ui.fragments.mainactivity.AbsQueueFragment
import com.o4x.musical.ui.viewmodel.HomeHeaderViewModel
import com.o4x.musical.ui.viewmodel.ScrollPositionViewModel
import com.o4x.musical.util.CoverUtil
import com.o4x.musical.util.MusicUtil
import com.o4x.musical.util.Util
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.koin.core.parameter.parametersOf
import org.koin.ext.scope
import java.io.File
import kotlin.math.max
import kotlin.math.min
import kotlin.properties.Delegates

class   HomeFragment : AbsQueueFragment(R.layout.fragment_home) {

    private val posterViewModel by sharedViewModel<HomeHeaderViewModel>()

    private val scrollPositionViewModel by viewModel<ScrollPositionViewModel>()

    private lateinit var recentlyAdapter: HomeAdapter
    private lateinit var newAdapter: HomeAdapter
    private lateinit var recentlyLayoutManager: GridLayoutManager
    private lateinit var newLayoutManager: GridLayoutManager


    // Heights //
    private var displayHeight by Delegates.notNull<Int>()
    private var appbarHeight by Delegates.notNull<Int>()
    private var toolbarHeight by Delegates.notNull<Int>()
    private var headerHeight by Delegates.notNull<Int>()

    private var _binding: FragmentHomeBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentHomeBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mainActivity.removeMusicServiceEventListener(posterViewModel)
        _binding = null
    }

    override fun onResume() {
        super.onResume()

        showAppbar()
        if (scrollPositionViewModel.getPositionValue() < headerHeight) {
            mainActivity.toolbar.setBackgroundColor(Color.TRANSPARENT)
            mainActivity.window.statusBarColor = Color.TRANSPARENT
            mainActivity.appbar.elevation = 0f
            setToolbarTitle(null)
        }

    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mainActivity.addMusicServiceEventListener(posterViewModel)
        setUpViews()
    }

    override fun showStatusBar() {}

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_home, menu)
        super.onCreateOptionsMenu(menu, inflater)
        val color = textColorTertiary()
        ToolbarContentTintHelper.colorizeToolbar(mainActivity.toolbar,
            color, serviceActivity)
        ToolbarContentTintHelper.tintAllIcons(menu, color)
        mainActivity.toggle.drawerArrowDrawable.color = color
        mainActivity.toggle.drawerArrowDrawable.alpha = 255
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_shuffle_all -> {
                libraryViewModel.shuffleSongs()
                return true
            }
            R.id.action_new_playlist -> {
                CreatePlaylistDialog.create().show(childFragmentManager, "CREATE_PLAYLIST")
                return true
            }
            R.id.action_search -> {
                mainActivity.openSearch()
                return true
            }
            R.id.action_reset_header -> {
                HomeHeaderPref.setDefault()
                return true
            }
            R.id.action_use_custom -> {
                startHomeHeaderImagePicker()
                return true
            }
            R.id.action_use_song -> {
                val myIntent = Intent(
                    requireContext(),
                    MusicPickerActivity::class.java
                )
                startActivityForResult(myIntent, REQUEST_CODE_SELECT_SONG)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setUpViews() {
        setUpHeights()
        setupButtons()
        setupPoster()
        setUpBounceScrollView()
        setUpRecentlyView()
        setUpNewView()
        setupEmpty()
    }


    private fun setUpHeights() {
        displayHeight = Util.getScreenHeight()
        var params: ViewGroup.LayoutParams

        // Set up header height
        params = binding.header.layoutParams
        params.height = displayHeight / 3
        binding.header.layoutParams = params

        // Set up poster image height
        params = binding.poster.layoutParams
        params.height = (displayHeight / 1.5f).toInt()
        binding.poster.layoutParams = params


        appbarHeight = appbarHeight()
        toolbarHeight = toolbarHeight()
        // get real header height
        headerHeight = binding.header.layoutParams.height - appbarHeight
    }

    private fun setupButtons() {
        binding.queueParent.setOnClickListener { mainActivity.setMusicChooser(R.id.nav_queue) }
        binding.queueShuffleButton.setOnClickListener {libraryViewModel.shuffleSongs() }
        binding.recentlyParent.setOnClickListener {
            navController.toPlaylistDetail(HistoryPlaylist())
        }
        binding.newlyParent.setOnClickListener {
            navController.toPlaylistDetail(LastAddedPlaylist())
        }


    }

    private fun setupPoster() {
        posterViewModel.getPosterBitmap().observe(viewLifecycleOwner, {
            posterViewModel.calculateBitmap(binding.poster, it)
        })
    }


    private fun setUpBounceScrollView() {

        var isStatusFlat = false
        var isAppbarFlat = false

        binding.nestedScrollView.setOnScrollChangeListener { _: NestedScrollView?, _: Int, scrollY: Int, _: Int, oldScrollY: Int ->

            scrollPositionViewModel.setPosition(scrollY)

            // Scroll poster
            binding.poster.y =
                ((-scrollY / (displayHeight * 2 / binding.poster.layoutParams.height.toFloat())).toInt())
                    .toFloat()

            // Scroll appbar
            if (scrollY > headerHeight + toolbarHeight && !isAppbarFlat) {
                setToolbarTitle(navController.currentDestination?.label.toString())
                toolbarColorVisible(true)
                mainActivity.appbar.elevation = resources.getDimension(R.dimen.appbar_elevation)
                isAppbarFlat = true
            }
            if (scrollY > headerHeight) { // outside header
                if (!isStatusFlat) {
                    statusBarColorVisible(true)
                    isStatusFlat = true
                }
                if (oldScrollY != 0) {
                    if (scrollY > oldScrollY) { // Scrolling up
                        mainActivity.appbar.y =
                            max(
                                -toolbarHeight.toFloat(),
                                mainActivity.appbar.y + (oldScrollY - scrollY)
                            )
                    } else { // Scrolling down
                        mainActivity.appbar.y = min(
                            0f, mainActivity.appbar.y + (oldScrollY - scrollY)
                        )
                    }
                }
            } else { // inside header
                if (isStatusFlat) {
                    setToolbarTitle(null)
                    toolbarColorVisible(false)
                    statusBarColorVisible(false)
                    mainActivity.appbar.elevation = 0f
                    isStatusFlat = false
                    isAppbarFlat = false
                }
                mainActivity.appbar.y = 0f
            }
        }

        // zooming poster in over scroll


        // zooming poster in over scroll
        val params = binding.poster.layoutParams
        val width = params.width
        val height = params.height
        binding.nestedScrollView.setOnOverScrollListener { _: Boolean, overScrolledDistance: Int ->
            val scale = 1 + overScrolledDistance / displayHeight.toFloat()
            val mParams: ViewGroup.LayoutParams =
                FrameLayout.LayoutParams(
                    width,
                    (height * scale).toInt()
                )
            binding.poster.layoutParams = mParams
        }
    }

    override fun initQueueView() {
        queueLayoutManager = GridHelper.linearLayoutManager(requireContext())
        binding.queueRecyclerView.layoutManager = queueLayoutManager
        queueAdapter = HomeAdapter(
            mainActivity,
            ArrayList(),
            0,
            R.layout.item_card_home,
            null,
            true
        )
        binding.queueRecyclerView.adapter = queueAdapter

        playerViewModel.queue.observe(viewLifecycleOwner, {
            binding.queueContainer.isVisible = it.isNotEmpty()
        })
    }

    private fun setUpRecentlyView() {
        recentlyLayoutManager = GridHelper.gridLayoutManager(requireContext())
        binding.recentlyRecyclerView.layoutManager = recentlyLayoutManager
        recentlyAdapter = HomeAdapter(
            mainActivity,
            ArrayList(),
            0,
            R.layout.item_card_home,
            requireContext().homeGridSize() * 2,
            false,
        )
        binding.recentlyRecyclerView.adapter = recentlyAdapter
        libraryViewModel.getRecentlyPlayed().observe(viewLifecycleOwner, {
            binding.recentlyContainer.isVisible = it.isNotEmpty()
            recentlyAdapter.swapDataSet(it)
        })
    }

    private fun setUpNewView() {
        newLayoutManager = GridHelper.gridLayoutManager(requireContext())
        binding.newRecyclerView.layoutManager = newLayoutManager
        newAdapter = HomeAdapter(
            mainActivity,
            ArrayList(),
            0,
            R.layout.item_card_home,
            requireContext().homeGridSize() * 3,
            false,
        )
        binding.newRecyclerView.adapter = newAdapter

        libraryViewModel.getRecentlyAdded().observe(viewLifecycleOwner, {
            binding.newlyContainer.isVisible = it.isNotEmpty()
            newAdapter.swapDataSet(it)
        })
    }


    private fun setupEmpty() {
        libraryViewModel.getSongs().observe(viewLifecycleOwner, {
            binding.empty.isVisible = it.isEmpty()
        })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            REQUEST_CODE_SELECT_IMAGE -> if (resultCode == Activity.RESULT_OK) {
                data?.data?.let { it ->
                    GlobalScope.launch(Dispatchers.IO) {
                        if (headerDir.isDirectory) {
                            val newImage = it.toFile()
                            headerDir.listFiles()?.let {
                                for (image in it) {
                                    if (image != newImage) image.delete()
                                }
                            }
                        }
                    }

                    HomeHeaderPref.customImagePath = it.toString()
                }
            }
            REQUEST_CODE_SELECT_SONG -> if (resultCode == Activity.RESULT_OK) {
                data?.data?.let {
                    HomeHeaderPref.imageSongID = MusicUtil.getSongIDFromFileUri(it)
                }
            }
        }
    }

    private val headerDir = File(App.getContext().filesDir, "/home_header/")

    private fun startHomeHeaderImagePicker() {
        ImagePicker.with(this)
            .saveDir(headerDir)
            .galleryOnly()
            .crop()
            .start(REQUEST_CODE_SELECT_IMAGE)
    }

    companion object {
        const val REQUEST_CODE_SELECT_IMAGE = 1600
        const val REQUEST_CODE_SELECT_SONG = 1700
    }
}